File: curvine-common/src/state/ttl_action.rs
// Copyright 2025 OPPO.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use num_enum::{FromPrimitive, IntoPrimitive};
use orpc::{err_box, CommonError};
use serde::{Deserialize, Serialize};

#[repr(i32)]
#[derive(
    Debug,
    Clone,
    Copy,
    Serialize,
    Deserialize,
    PartialEq,
    IntoPrimitive,
    FromPrimitive,
    Eq,
    Default,
    Hash,
)]
pub enum TtlAction {
    /// No action is performed.
    #[default]
    None = 0,

    /// Try moving to slow storage such as disk
    Move = 1,

    /// Move to the underlying storage after expiration
    Ufs = 2,

    /// Delete after expiration
    Delete = 3,
}

impl TryFrom<&str> for TtlAction {
    type Error = CommonError;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        let action = match value.to_uppercase().as_str() {
            "NONE" => TtlAction::None,
            "MOVE" => TtlAction::Move,
            "UFS" => TtlAction::Ufs,
            "DELETE" => TtlAction::Delete,

            _ => return err_box!("invalid ttl action: {}", value),
        };

        Ok(action)
    }
}

File: curvine-libsdk/java/src/main/java/io/curvine/TtlAction.java
// Copyright 2025 OPPO.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package io.curvine;

import io.curvine.proto.TtlActionProto;

/**
 * TTL动作枚举，对应Rust中的TtlAction
 */
public enum TtlAction {
    NONE(0),
    FREE(1),
    DELETE(2);

    private final int value;

    TtlAction(int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }

    /**
     * 从Protobuf枚举转换
     */
    public static TtlAction fromProto(TtlActionProto proto) {
        switch (proto) {
            case TTL_ACTION_PROTO_NONE:
                return NONE;
            case TTL_ACTION_PROTO_UFS:
                return FREE;
            case TTL_ACTION_PROTO_DELETE:
                return DELETE;
            default:
                throw new IllegalArgumentException("Unknown TtlActionProto: " + proto);
        }
    }

    /**
     * 转换为Protobuf枚举
     */
    public TtlActionProto toProto() {
        switch (this) {
            case NONE:
                return TtlActionProto.TTL_ACTION_PROTO_NONE;
            case FREE:
                return TtlActionProto.TTL_ACTION_PROTO_UFS;
            case DELETE:
                return TtlActionProto.TTL_ACTION_PROTO_DELETE;
            default:
                throw new IllegalArgumentException("Unknown TtlAction: " + this);
        }
    }
}
